# Recursion Assignment

# Solve the recursion
f(n) = 7f(n-1) - 10f(n-2)  f(0)=1 f(1)=3

1. Using mathematical method taught in class to get analytic equation. Type up details and program function

In [1]:
# Analytic solution of the given recurrence relation:
# f(n) = 7f(n-1) - 10f(n-2)
# Given: f(0) = 1, f(1) = 3

# Step 1: Assume f(n) = r^n
# Substitute in the relation → r^2 - 7r + 10 = 0
# Solving gives (r - 5)(r - 2) = 0 → r1 = 5, r2 = 2

# Step 2: General solution
# f(n) = A*(5^n) + B*(2^n)

# Step 3: Using initial conditions
# when n=0 → A + B = 1
# when n=1 → 5A + 2B = 3
# Solving gives A = 1/3 and B = 2/3

# Final analytic formula:
# f(n) = (1/3)*5^n + (2/3)*2^n
# To keep results exact (integer form): f(n) = (5^n + 2*(2^n)) / 3

def f_analytic(n):
    if n < 0:
        print("n must be non-negative")
        return None
    result = (pow(5, n) + 2 * pow(2, n)) // 3
    return result

# Checking first few terms to verify
for i in range(6):
    print("f({}) = {}".format(i, f_analytic(i)))


f(0) = 1
f(1) = 3
f(2) = 11
f(3) = 47
f(4) = 219
f(5) = 1063


2. Using dynamic programming.

In [2]:
# Dynamic Programming approach (Iterative / Bottom-Up)
# Given: f(n) = 7f(n-1) - 10f(n-2)
# f(0) = 1, f(1) = 3

def f_dp_iter(n):
    if n < 0:
        print("n must be non-negative")
        return None
    if n == 0:
        return 1
    if n == 1:
        return 3
    
    f0, f1 = 1, 3
    for i in range(2, n + 1):
        f_next = 7 * f1 - 10 * f0
        f0, f1 = f1, f_next
    return f1

# checking few values to make sure it's working
for i in range(6):
    print("f({}) = {}".format(i, f_dp_iter(i)))


f(0) = 1
f(1) = 3
f(2) = 11
f(3) = 47
f(4) = 219
f(5) = 1063


3. Using dynamic programming and memoization.

In [3]:
# Dynamic Programming with Memoization (Top-Down)
# Here we use recursion with caching to avoid repeated calculations

from functools import lru_cache

@lru_cache(maxsize=None)
def f_dp_memo(n):
    if n < 0:
        print("n must be non-negative")
        return None
    if n == 0:
        return 1
    if n == 1:
        return 3
    return 7 * f_dp_memo(n - 1) - 10 * f_dp_memo(n - 2)

# check few values
for i in range(6):
    print("f({}) = {}".format(i, f_dp_memo(i)))


f(0) = 1
f(1) = 3
f(2) = 11
f(3) = 47
f(4) = 219
f(5) = 1063


4. Compare the results of each function. Discuss the advantages and disadvantages of each method. Consider difficulty of solving and the time it would take to get the 1000 entry of the sequence in each of the three solutions.

After running all three methods;  analytic, dynamic programming (iterative), and dynamic programming with memoization, the outputs are the same for every value of *n*.  
This proves that all three approaches correctly solve the given recurrence relation.

**Observation:**  
For small values like f(2), f(3), and f(4), all three give 11, 47, and 219 respectively.  
When I tested f(1000), every method gave the same huge number (only the analytic one was almost instant).



#### 1. Analytic Method
The analytic or closed-form solution uses the formula  
**f(n) = (1/3)×5ⁿ + (2/3)×2ⁿ**  
This is the fastest way because it calculates the answer directly using powers.  
However, the formula takes more time to derive at the beginning since we need to solve the characteristic equation first.


#### 2. Dynamic Programming (Iterative)
This method builds the result step by step from f(0) and f(1).  
It is easy to understand and very reliable.  
It takes O(n) time and O(1) space, so even f(1000) can be found in less than a second.  
This is usually the best choice for programming tasks.


#### 3. Dynamic Programming with Memoization
This is the recursive approach that stores already-calculated results.  
It gives the same answers but is a bit slower than the iterative one because of recursion overhead.  
It is good for understanding how recursion works but not always the most efficient for very large *n*.



#### Final Comparison
| Method | Type | Time | Space | Remarks |
| Analytic | Formula / Math | Fastest (O(log n)) | O(1) | Needs derivation but gives instant results |
| DP Iterative | Bottom-Up | O(n) | O(1) | Simple, clean, and efficient |
| DP Memoized | Top-Down | O(n) | O(n) | Easy to write but uses recursion and more memory |

##Conclusion 
All three methods are correct.  
The analytic formula is the fastest, the iterative DP is the easiest to code, and the memoized version is good for learning the concept.  
For calculating something like f(1000), the analytic one is almost instant, while the DP ones are still quite fast.
